include/range/v3/detail/optional.hpp

```diff
@@ -86,6 +86,49 @@ namespace ranges
             }
         };
 
+        template<typename T>
+        struct optional<T &>
+        {
+        private:
+            variant<meta::nil_, T &> data_;
```

Comment:
This would be more efficiently implemented atop a `T*` than a variant, using `null` to represent the empty state.